### PR TITLE
rename: MathConst => Irrational

### PR DESCRIPTION
**Pull request by StefanKarpinski**
I have to say this feels like a _much_ better name:

``` jl
julia> typeof(π)
Irrational{:π}
```

### CODE REVIEW

**Comment by quinnj** [comment]
+1

**Comment by tkelman** [comment]
I like. #11200 says hello though.

**Comment by StefanKarpinski** [comment]
Right, what we really need is a way to deprecate exported bindings. Bonus: don't autocomplete those or otherwise expose them as suggestions.

File: base/irrationals.jl [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        @assert Float64($esym) == Float64(big($esym))
+        @assert Float32($esym) == Float32(big($esym))
+    end
+end
+
+big(x::Irrational) = convert(BigFloat,x)
+
+## specific mathematical constants
+
+@math_const π        3.14159265358979323846  pi
```
Comment (waldyrious) [inline base/irrationals.jl]: Should `@math_const` change its name too?

File: base/irrationals.jl [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+@generated function call{T<:Union{Float32,Float64},s}(t::Type{T},c::Irrational{s},r::RoundingMode)
+    f = T(big(c()),r())
+    :($f)
+end
+
+=={s}(::Irrational{s}, ::Irrational{s}) = true
+==(::Irrational, ::Irrational) = false
+
+# MathConsts are irrational, so unequal to everything else
```
Comment (waldyrious) [inline base/irrationals.jl]: Slightly redundant comment / needs rephrasing

File: base/irrationals.jl [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        Base.convert(::Type{Float32}, ::Irrational{$qsym}) = $(Float32(val))
+        @assert isa(big($esym), BigFloat)
+        @assert Float64($esym) == Float64(big($esym))
+        @assert Float32($esym) == Float32(big($esym))
+    end
+end
+
+big(x::Irrational) = convert(BigFloat,x)
+
+## specific mathematical constants
```
Comment (waldyrious) [inline base/irrationals.jl]: specify "irrational" here too?

**Comment by StefanKarpinski** [comment]
Thanks for the feedback, @waldyrious!

**Comment by waldyrious** [comment]
Glad I can help :)

**Comment by dpsanders** [comment]
+1

File: base/irrationals.jl [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+@generated function call{T<:Union{Float32,Float64},s}(t::Type{T},c::Irrational{s},r::RoundingMode)
+    f = T(big(c()),r())
+    :($f)
+end
+
+=={s}(::Irrational{s}, ::Irrational{s}) = true
+==(::Irrational, ::Irrational) = false
+
+# Irationals are not rational, so unequal to everything else
```
Comment (waldyrious) [inline base/irrationals.jl]: Typo: Irationals is missing an "r". Also, I'm not sure I fully understand this comment. The beginning is redundant and the ending is kinda vague.

**Comment by GaborOszlanyi** [comment]
-1
Just an other opinion: ```MathConst``` was a better name.

**Comment by simonbyrne** [comment]
Now we have generated functions, we could also get rid of the numeric argument to the macro: since we're renaming, perhaps this would be a good time to make the switch?

**Comment by simonbyrne** [comment]
Also, I like the change: it helps make clear why `pi != float(pi)`

**Comment by ivarne** [comment]
I like this change, but 5 hours seems like a too short comment period, especially with a global audience.

**Comment by ivarne** [comment]
I like this change, but 5 hours seems like a too short comment period, especially with a global audience.

**Comment by StefanKarpinski** [comment]
Yeah, sorry about that. Got a bit trigger happy.

**Comment by simleb** [comment]
I'm not convinced. It might look nicer to the eye but then why wouldn't `π+1` be promoted to `Irrational` as well? What about `√2`?

What I'm saying is that using the word irrational, which has a well understood mathematical meaning, for a type that wraps mathematical constants which happen to be irrational but are just a tiny subset of all irrational numbers, especially when the language already has a `Rational` type with a widely different use (integers are promoted to rational for instance…), is confusing.

I know it's just a name, and the type of these mathematical constants is meant to be transient (in the sense that they almost always appear as part of expressions which convert them) but it is confusing I think.

**tl;dr** What is `Irrational`? The type used for mathematical constants… obviously!

**Comment by simonbyrne** [comment]
@simleb Good point. What about `IrrationalConst`?

**Comment by dpsanders** [comment]
+1 for `IrrationalConst`. (I was just typing almost the same.)

**Comment by ScottPJones** [comment]
:+1: also for `IrrationalConst`

**Comment by GaborOszlanyi** [comment]
We had a good, obvious and short name for this.
Why not leave it as is?

**Comment by StefanKarpinski** [comment]
`MathConst` was far from obvious – there was a fair amount of confusion about it over time and it always bothered me. To argue in favor of `Irrational` rather than `IrrationalConst`, this type is capable of representing arbitrary irrational values as long as you have a way to define them. It's not like anyone is going to come up with a clever encoding that can represent all irrational numbers.

**Comment by waldyrious** [comment]
I think `IrrationalConst` may be clearer since (in computing) using the term "constants" specifies that these are hand-picked numbers rather than a type meant to represent any irrational number.

File: base/irrationals.jl [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        Base.convert(::Type{Float32}, ::Irrational{$qsym}) = $(Float32(val))
+        @assert isa(big($esym), BigFloat)
+        @assert Float64($esym) == Float64(big($esym))
+        @assert Float32($esym) == Float32(big($esym))
+    end
+end
+
+big(x::Irrational) = convert(BigFloat,x)
+
+## specific irriational mathematical constants
```
Comment (lostanlen) [inline base/irrationals.jl]: typo here: "irriational" -> "irrational".

**Comment by simleb** [comment]
Well, the name `Irrational` is starting to grow on me. I agree with the arguments set forth by Stefan in https://github.com/JuliaLang/julia/pull/11929: `Irrational` simply couldn't mean anything else in the context of Julia. There is no way to even test if a number is irrational so there is no way to create them outside of how it is done now: by providing an algorithm to compute them to arbitrary precision. And since these numbers clearly have to be constant, there is no need for `Const` in the name. (Plus there is really no advantage using `IrrationalConst` over `MathConst`.)

Thus from now on I fully support the name `Irrational` for these mathematical constants (until someone writes a clever rebuttal, haha)!
